tests/tracer/test_single_span_sampling_rule.py

```diff
@@ -0,0 +1,149 @@
+
+from ..utils import DummyTracer
+
+
+def assert_sampling_decision_tags(span, sample_rate=1.0, mechanism=SamplingMechanism.SPAN_SAMPLING_RULE, limit=None):
+    assert span.get_metric(_SINGLE_SPAN_SAMPLING_RATE) == sample_rate
+    assert span.get_metric(_SINGLE_SPAN_SAMPLING_MECHANISM) == mechanism
+    assert span.get_metric(_SINGLE_SPAN_SAMPLING_MAX_PER_SEC) == limit
+
+
+def traced_function(rule, name="test_name", service="test_service"):
+    tracer = DummyTracer()
+    with tracer.trace(name) as span:
+        span.service = service
+        rule.sample(span)
+    return span
+
+
+def test_single_span_rule_default_no_match():
+    rule = SpanSamplingRule()
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span, sample_rate=None, mechanism=None, limit=None)
+
+
+def test_single_span_rule_no_match_service():
+    rule = SpanSamplingRule(service="wrong_service_name", name="test_name")
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span, sample_rate=None, mechanism=None, limit=None)
+
+
+def test_single_span_rule_no_match_name():
+    rule = SpanSamplingRule(service="test_service", name="wrong_operation_name")
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span, sample_rate=None, mechanism=None, limit=None)
+
+
+def test_single_span_rule_no_match_only_service():
+    rule = SpanSamplingRule(service="wrong_service_name")
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span, sample_rate=None, mechanism=None, limit=None)
+
+
+def test_single_span_rule_no_match_only_name():
+    rule = SpanSamplingRule(name="wrong_operation_name")
+    span = traced_function(rule)
+    assert_sampling_decision_tags(span, sample_rate=None, mechanism=None, limit=None)
+
+
+def test_single_span_rule_match():
+    rule = SpanSamplingRule(service="test_service", name="test_name")
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span)
+
+
+def test_single_span_rule_match_only_service():
+    rule = SpanSamplingRule(service="test_service")
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span)
+
+
+def test_single_span_rule_match_only_name():
+    rule = SpanSamplingRule(name="test_name")
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span)
+
+
+# More extensive testing of the matching is done in test_glob_matcher.py
+def test_single_span_rule_match_wildcards():
+    rule = SpanSamplingRule(service="test_*", name="test?????")
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span)
+
+
+def test_single_span_rule_no_match_wildcards():
+    rule = SpanSamplingRule(service="*test_", name="test_nam??")
+    span = traced_function(rule)
+
+    assert_sampling_decision_tags(span, sample_rate=None, mechanism=None, limit=None)
+
+
+def test_single_span_rule_unsupported_pattern_bracket_expansion():
+    rule = SpanSamplingRule(service="test_servic[a-z]+", name="test_name")
```

Comment:
Should we log a warning? or raise an exception?

Reply:
We raise exceptions in the sampler code at various points because we want misconfigured rules to be surfaced in development. Thinking forward to when we will be reading these span sampling rules from the environment, we'll want to follow the same practice.

Reply:
Discussed and decided that for now we won't try to throw an exception or otherwise when a customer uses regex. Instead we'll just evaluate it as a string.